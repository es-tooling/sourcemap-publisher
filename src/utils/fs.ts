import {fdir} from 'fdir';
import path from 'node:path';
import {mkdir, rm, stat, cp} from 'node:fs/promises';

export async function getSourceFilesFromPaths(
  cwd: string,
  paths: string[]
): Promise<string[]> {
  const crawler = new fdir();
  const files = await crawler
    .withFullPaths()
    .exclude((_dirName, dirPath) => {
      return !paths.some((p) => dirPath.startsWith(p));
    })
    .filter((file) => {
      return (
        paths.some((p) => file.startsWith(p)) &&
        (file.endsWith('.js') ||
          (!file.endsWith('.d.ts') && file.endsWith('.ts')))
      );
    })
    .crawl(cwd)
    .withPromise();
  return files;
}

export async function getTempDir(cwd: string, name: string): Promise<string> {
  const tempDir = path.join(cwd, name);

  await rm(tempDir, {force: true, recursive: true});
  await mkdir(tempDir);

  return tempDir;
}

export async function copyRelativeFilesToDir(
  files: string[],
  sourceDir: string,
  targetDir: string
): Promise<void> {
  for (const file of files) {
    const sourcePath = path.join(sourceDir, file);
    const targetPath = path.join(targetDir, file);
    try {
      await stat(sourcePath);
      await cp(sourcePath, targetPath, {recursive: true});
    } catch {
      continue;
    }
  }
}
