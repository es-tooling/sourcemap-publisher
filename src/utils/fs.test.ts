import {suite, beforeEach, afterEach, test, expect} from 'vitest';
import {writeFile, rm, mkdtemp, mkdir, stat} from 'node:fs/promises';
import {copyRelativeFilesToDir, getTempDir} from './fs.js';
import path from 'node:path';
import {tmpdir} from 'node:os';

const mockFs: Record<string, string> = {
  'lib/js-file.js': '// foo',
  'lib/ts-file.ts': '// foo',
  'lib/dts-file.d.ts': '// foo',
  'lib/nested/js-file.js': '// foo'
};

const writeMockFs = async (tempDir: string) => {
  for (const [file, content] of Object.entries(mockFs)) {
    const filePath = path.join(tempDir, file);
    await mkdir(path.dirname(filePath), {recursive: true});
    await writeFile(filePath, content);
  }
};

suite('getTempDir', () => {
  let tempDir: string;

  beforeEach(async () => {
    tempDir = await mkdtemp(path.join(tmpdir(), 'smpub'));
  });

  afterEach(async () => {
    await rm(tempDir, {force: true, recursive: true});
  });

  test('creates temporary directory', async () => {
    const result = await getTempDir(tempDir, 'woof');
    const expected = path.join(tempDir, 'woof');
    expect(result).to.equal(expected);
    await expect(stat(expected)).resolves.not.toThrow();
  });
});

suite('copyRelativeFilesToDir', () => {
  let tempDir: string;
  let targetDir: string;

  beforeEach(async () => {
    tempDir = await mkdtemp(path.join(tmpdir(), 'smpub'));
    targetDir = await mkdtemp(path.join(tmpdir(), 'smpub'));
    await writeMockFs(tempDir);
  });

  afterEach(async () => {
    await rm(tempDir, {force: true, recursive: true});
    await rm(targetDir, {force: true, recursive: true});
  });

  test('copies files to target directory', async () => {
    const files = ['lib/js-file.js', 'lib/ts-file.ts'];
    await copyRelativeFilesToDir(files, tempDir, targetDir);

    await expect(
      stat(path.join(targetDir, 'lib/js-file.js'))
    ).resolves.not.toThrow();
    await expect(
      stat(path.join(targetDir, 'lib/ts-file.ts'))
    ).resolves.not.toThrow();
  });

  test('ignores non-existent files', async () => {
    const files = ['lib/js-file.js', 'lib/non-existent-file.js'];
    await copyRelativeFilesToDir(files, tempDir, targetDir);

    await expect(
      stat(path.join(targetDir, 'lib/js-file.js'))
    ).resolves.not.toThrow();
    await expect(
      stat(path.join(targetDir, 'lib/non-existent-file.js'))
    ).rejects.toThrow();
  });
});
