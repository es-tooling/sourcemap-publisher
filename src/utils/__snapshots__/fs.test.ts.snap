// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getSourceFilesFromPaths > should find all js/ts files in the specified paths 1`] = `
[
  "<TEMPDIR>/lib/js-file.js",
  "<TEMPDIR>/lib/nested/js-file.js",
]
`;
